### PR TITLE
fix(hook): resolve cargo package name from Cargo.toml, not dir basename (#4512)

### LINKED ISSUE #4512: fix(hook): pre-push cargo fmt uses dir basename instead of Cargo.toml package name
## Summary

Pre-push hook (`cargo xtask fmt --check -p <name>`) infers the package name from the crate directory basename, which is wrong for `crates/perl-lsp/` (package name is `perl-lsp-rs`, not `perl-lsp`). The command fails, the hook rejects the push, and builders bypass with `--no-verify` — losing the fmt-check guarantee.

## Evidence

Observed 2× on 2026-04-19 collapse session:
- G1a builder push (commit series)
- G1b builder fix push (commit `d9ecd697c`, PR #4510)

Both cited the same reason: `cargo fmt -p perl-lsp` fails because no package named `perl-lsp` exists in the workspace.

## Impact

- Pre-push fmt gate silently bypassed — relies on CI to catch fmt issues instead of local.
- Trains agents to reach for `--no-verify` when the hook misbehaves, which is exactly the pattern we warn against in memory (`feedback_pre_push_hook_windows_race.md`).
- Root cause conflated with Windows MAX_PATH complaints — muddies diagnosis of the real MAX_PATH bug (#4456).

## Root cause

Hook uses directory basename as the package name. For `crates/perl-lsp/` the actual package name is `perl-lsp-rs`. Every other crate in the workspace has matching dir and package name, so the hook works elsewhere — masking the bug.

## Fix

Replace basename inference with `cargo metadata --no-deps --format-version 1` lookup:

```bash
# Pseudo-Bash / adjust for the actual hook language
# Old:
PKG_NAME="$(basename "$CRATE_DIR")"

# New:
PKG_NAME="$(cargo metadata --no-deps --format-version 1 \
  | jq -r --arg dir "$CRATE_DIR" \
    '.packages[] | select(.manifest_path | startswith($dir + "/")) | .name' \
  | head -1)"

# Fallback if cargo metadata not available (shouldn't happen):
[[ -z "$PKG_NAME" ]] && PKG_NAME="$(basename "$CRATE_DIR")"
```

Implementation location: the Rust-based hook runner in `xtask/src/tasks/fmt.rs` (or wherever `xtask fmt` resolves package names). Use `cargo_metadata` crate directly rather than shelling out.

## Acceptance

- Pre-push hook on a modification to `crates/perl-lsp/src/foo.rs` invokes `cargo fmt --check -p perl-lsp-rs` (not `-p perl-lsp`)
- Hook succeeds when formatting is clean
- No `--no-verify` needed for routine `perl-lsp-rs` edits
- Regression test in `xtask/tests/` using a synthetic workspace fixture where dir and package name differ

## Out of scope

- Renaming `crates/perl-lsp/` to match package name — separate issue #4511, scheduled after G1b
- Fixing other `--no-verify` patterns caused by different bugs (MAX_PATH, hook race) — tracked separately

## Alternatives considered

1. **Rename the crate directory instead** (tracked as #4511). Fixes THIS crate but not the underlying brittleness. The hook will break again if any future crate has dir/package name mismatch.
2. **Document `--no-verify` as acceptable**. No — trains against the memory rule `feedback_pre_push_hook_windows_race.md`.
3. **This fix** — read Cargo.toml via cargo_metadata. Best; also works for future crates.

## Pipeline entry

`builder-ready` — small mechanical fix. Skip verification pipeline (nothing external to verify; the change is tooling-internal).

**Relates to:** #4511 (rename complementary fix), #4509 (task-tool persistence — unrelated hook).

### PR DESCRIPTION
**Pull request by EffortlessSteven**
## Summary

- Pre-push single-crate gate used directory basename as the `-p` package name argument, causing `cargo fmt -p perl-lsp` to fail because the actual package name is `perl-lsp-rs`
- Adds `cargo xtask resolve-package-name <crate-dir>` subcommand that looks up the true package name from `cargo metadata`
- Fixes `hooks/pre-push` to call the new subcommand; falls back to dir basename if xtask is unavailable

## Changes

- `xtask/src/tasks/targeted_checks.rs` — make `resolve_package_names` public; add `resolve_single_package_name(root, crate_dir) -> Result<String>`; fix Windows path separator normalization (`\` → `/`) in manifest path matching
- `xtask/src/main.rs` — add `ResolvePackageName { crate_dir }` CLI subcommand; uses `current_dir()` as workspace root for testability in synthetic workspaces
- `hooks/pre-push` — replace `SINGLE_CRATE_NAME="$(... | tr -d '[:space:]')"` (basename) with `cargo xtask resolve-package-name "crates/${SINGLE_CRATE_DIR}"` (Cargo.toml lookup) with dirname fallback
- `xtask/tests/fmt_package_lookup_tests.rs` — 6 integration tests covering: subcommand help exists, dir≠pkg name (synthetic), real workspace (`crates/perl-lsp` → `perl-lsp-rs`), dir=pkg name (normal case), unknown dir errors, single clean line output

## Evidence

- **Commits**: 3 on branch (spec, red tests, implementation)
- **Tests**: 6/6 pass (`cargo test -p xtask --test fmt_package_lookup_tests`)
- **Lint**: 0 new warnings (`cargo clippy -p xtask --tests`)
- **Files**: 4 production/test files changed, +62/-8 lines (excluding spec files)
- **PR-fast gate**: passed on push

## Test Plan

- `cargo test -p xtask --test fmt_package_lookup_tests` — all 6 tests pass
- `cargo xtask resolve-package-name crates/perl-lsp` from workspace root — outputs `perl-lsp-rs`
- Push a change to `crates/perl-lsp/src/` — hook runs `cargo fmt -p perl-lsp-rs` (not `-p perl-lsp`), succeeds when formatting is clean
- No `--no-verify` needed for routine `perl-lsp-rs` edits

## Unknowns

- `.git/hooks/pre-push` is synced manually (cp) in this PR; the self-heal logic (lines 44-50 of the hook) will keep future installs in sync automatically
- The `resolve-package-name` subcommand uses `current_dir()` not `project_root()` — this means it works in any Cargo workspace, not just perl-lsp. Intentional.
- Pre-existing 2 failing tests in `check_test_wiring` are unrelated to this change (confirmed: `check_test_wiring.rs` has 0 lines changed on this branch)

Closes #4512

### CODE REVIEW

**Comment by EffortlessSteven** [comment]
## Reviewer: Standards Check

**Branch:** \`impl/4512-fmt-package-lookup\` @ \`8be02c9ff\`

### Scope
Clean. Diff touches exactly the 4 files in spec scope + \`.spec/\` history files. No drift.

### Banned patterns
- Production code (\`xtask/src/\`): no \`unwrap()\`, \`expect()\`, \`panic!\`, \`todo!\`, \`dbg!\`, \`.get(0)\`. Clean.
- Test helpers use \`.expect("..."\))\` with descriptive messages — same pattern as \`post_merge_status_regen_tests.rs\`. Acceptable for fixture setup.
- All 12 test functions use \`Result<()>\` return with \`?\` — correct.
- All \`assert!\` calls include a message string — correct.

### Formatting
\`cargo fmt --check\` passes. No changes needed.

### Clippy
xtask binary: clean (\`cargo clippy --manifest-path xtask/Cargo.toml -- -D warnings\`). The pre-existing \`len_zero\` warning in \`wave1_perl_module_collapse_tests.rs:436\` is not in this diff and was present on master before this branch.

### Changes made
None — fix forward had nothing to fix. PR was already clean.

### Verdict
Standards pass. Routing to reviewer-deep (required, every PR).

---
*Reviewer (standards) — no structural issues found.*

**Comment by EffortlessSteven** [comment]
## Maintainer PR: Project Fit Check

**Verdict: ALIGNED**

### Scope discipline
Diff matches spec exactly. 5 files: \`xtask/src/main.rs\`, \`xtask/src/tasks/targeted_checks.rs\`, \`hooks/pre-push\`, \`xtask/tests/fmt_package_lookup_tests.rs\`, and \`.spec/\` history files. No unrequested changes.

### Pattern introduction
New pattern: \`resolve_single_package_name\` as a public helper on \`targeted_checks\` + a new xtask CLI subcommand. This follows the existing convention in \`targeted_checks.rs\` (two other public functions: \`resolve_package_names\`, \`run\`). The CLI dispatch pattern matches every other subcommand in \`main.rs\`. No novel pattern introduced.

### Complexity budget
+36 production lines (main.rs + targeted_checks.rs + hook), ~400 test lines. Appropriate for a bug fix that unblocks the entire \`perl-lsp\` editing workflow on pre-push. High value-to-complexity ratio.

### Consistency
Error handling uses \`eyre!\` and \`?\` — matches the rest of xtask. Windows path normalization (\`replace('\\', "/")\`) is consistent with the repo's cross-platform approach. Hook fallback preserves old behavior if xtask is unavailable — good defensive design.

### Test quality
12 integration tests: happy path (mismatch, match, real workspace), error path (unknown dir, empty, dot), boundary (trailing slash, triple slash, stdout cleanliness, Windows compat). Tests the right thing.

### Documentation debt
\`features.toml\` is LSP-only — xtask internal commands don't register there. \`CLAUDE.md\` quick-reference doesn't list every xtask subcommand (none of the others are listed either). No documentation debt.

### Backwards compatibility
Hook fallback (\`|| printf '%s' "$SINGLE_CRATE_DIR"\`) preserves old behavior when xtask is unavailable. No breaking change.

Proceed to reviewer-deep.

---
*Maintainer PR — ALIGNED.*

File: xtask/tests/fmt_package_lookup_tests.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    let dir = TempDir::new().expect("create tempdir");
+    let root = dir.path();
+    fs::write(
+        root.join("Cargo.toml"),
+        r#"[workspace]
+members = []
+resolver = "2"
+"#,
+    )
+    .expect("write workspace Cargo.toml");
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply xtask/tests/fmt_package_lookup_tests.rs]: Applied. Fixture helpers now return `Result<TempDir>` with `?` propagation. All 12 tests pass. Commit `5227a03fa`.

File: .spec/4512-fmt-package-lookup/checklist.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+Files OUT of scope:
+- `xtask/src/tasks/fmt.rs` -- already correct
+- `.git/hooks/pre-push` -- auto-updated from hooks/pre-push
+- `Cargo.toml` / `xtask/Cargo.toml` -- no new runtime deps needed
+- `crates/perl-lsp/` rename -- tracked as #4511
+
+## Flags for builder
+
+1. **Check lib.rs first**: Run `ls xtask/src/lib.rs`. If absent, create minimal one before writing tests.
+2. **Windows path separator**: `resolve_package_names` uses `strip_prefix('/')`. On Windows manifest paths may use backslash. Check the existing function for Windows compat; normalize with `.replace('\', '/')` if needed before stripping.
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply .spec/4512-fmt-package-lookup/checklist.md]: Fixed in `.spec/4512-fmt-package-lookup/checklist.md` — the example now uses `.replace('\\', '/')`. Commit `5227a03fa`.

File: xtask/src/main.rs [hunk 3]
```diff
@@ -1561,6 +1572,15 @@ fn main() -> Result<()> {
             verbose,
         }),
         Commands::TargetedChecks { base, mode } => targeted_checks::run(base, mode),
+        Commands::ResolvePackageName { crate_dir } => {
+            // Use the current working directory as workspace root so this subcommand
+            // works correctly both in the main workspace and in test synthetic workspaces.
+            let root = std::env::current_dir()
+                .map_err(|e| eyre!("Failed to get current working directory: {e}"))?;
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply xtask/src/main.rs:1579]: This is intentional per `.spec/4512-fmt-package-lookup/context.md`. The subcommand uses `current_dir()` so the integration test suite can point it at synthetic workspaces via `.current_dir(ws.path())`. The hook always invokes `cargo xtask` from the git repo root (git hooks always run from the worktree root), so `current_dir()` == project root in every real invocation. Using `utils::project_root()` (which is baked at compile-time via `env!("CARGO_MANIFEST_DIR")`) would make synthetic-workspace testing impossible without patching the binary.

File: xtask/src/tasks/targeted_checks.rs [hunk 4]
```diff
@@ -137,8 +137,11 @@ fn resolve_package_names(
...
         // e.g., "/path/to/project/crates/perl-parser/Cargo.toml" -> "crates/perl-parser"
-        let relative = manifest_path
-            .strip_prefix(root_str.as_ref())
+        // Normalize separators to forward slashes for cross-platform compatibility (Windows uses backslash).
+        let manifest_normalized = manifest_path.replace('\\', "/");
+        let root_normalized = root_str.replace('\\', "/");
+        let relative = manifest_normalized
+            .strip_prefix(root_normalized.as_str())
             .and_then(|p| p.strip_prefix('/'))
             .and_then(|p| p.strip_suffix("/Cargo.toml"));
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply xtask/src/tasks/targeted_checks.rs:146]: The explicit `replace('\\', '/')` was added specifically to fix the Windows path separator regression (one of the two bugs this PR addresses). The `Path::strip_prefix` suggestion would bypass the normalization step, re-introducing the Windows failure mode. The manual approach is intentional. The test `resolve_windows_path_separator_compat_via_real_workspace` runs on Windows CI and guards this code path.

**Comment by EffortlessSteven** [comment]
## Refactor Plan

**Analysis of diff on branch \`impl/4512-fmt-package-lookup\` @ \`5227a03fa\`**

### 1. Minor style: `println!("{}", name)` → `println!("{name}")`
- **Where:** \`xtask/src/main.rs:1581\`
- **What:** Edition 2024 workspace — captured format string identifiers are idiomatic
- **How:** \`println!("{name}");\`
- **Risk:** Cosmetic only, no behavior change

### 2. Visibility: `resolve_package_names` could be `pub(crate)`
- **Where:** \`xtask/src/tasks/targeted_checks.rs:102\`
- **What:** \`resolve_package_names\` is only called from within \`targeted_checks.rs\` (at line 279 via \`resolve_single_package_name\`). \`resolve_single_package_name\` is the external API surface. Making \`resolve_package_names\` \`pub\` exposes implementation detail.
- **How:** Change \`pub fn resolve_package_names\` to \`pub(crate) fn resolve_package_names\`
- **Constraint:** xtask has no lib.rs target, so \`pub\` vs \`pub(crate)\` has no practical difference for external consumers. This is style, not a real encapsulation win. **Mark as skip unless green-refactor considers it worth the noise.**

### Skip (not worth changing)
- **Type tightness \`crate_dir: String\` in CLI struct:** Clap derives require owned \`String\` for positional args. Can't be \`&str\`. Correct as-is.
- **Manual path normalization in `resolve_package_names`:** The \`replace('\\', "/")\` approach is intentional (Windows fix). The bot's \`Path::strip_prefix\` suggestion would regress Windows. Keep as-is.
- **\`current_dir()\` in dispatch:** Design decision explained in context.md. Keep as-is.
- **Fixture helper duplication in tests:** Three \`make_*_workspace\` helpers share a common shape but vary in content — extracting a common builder would obscure the fixture intent. Keep as-is.

### Recommended for green-refactor
Only item 1 is clearly worth making: the \`println!("{name}")\` style improvement. Item 2 is borderline — do it only if green-refactor agrees it reduces surface without adding noise.

---
*Refactor planner — analysis only, no edits made.*

**Comment by EffortlessSteven** [comment]
## Green Refactor: Executed

**Commit:** \`6059c466c\`

### Changes made
1. **\`xtask/src/main.rs:1581\`** — \`println!("{}", name)\` → \`println!("{name}")\` — edition 2024 captured identifier style
2. **\`xtask/src/tasks/targeted_checks.rs:102\`** — \`pub fn resolve_package_names\` → \`pub(crate) fn resolve_package_names\` — tighter visibility; only called from within the same file

### Skipped (from refactor plan)
- \`crate_dir: String\` in CLI struct — Clap requires owned String for positional args, can't be \`&str\`
- Manual path normalization — intentional Windows fix, keep as-is
- \`current_dir()\` design — intentional for test testability, keep as-is
- Fixture helper deduplication — structural variation in content, keep readable as-is

**All 12 tests pass. Behavior unchanged.**

---
*Green refactor — behavior-preserving, tests stay green.*

**Comment by EffortlessSteven** [comment]
## Reviewer Deep: Correctness Check

**Branch:** \`impl/4512-fmt-package-lookup\` @ \`6059c466c\`

### What I verified

**Hook correctness:**
- \`SINGLE_CRATE_NAMES\` stores bare crate directory names (e.g., \`"perl-lsp"\`), confirmed by reading the extraction logic at lines ~115-125
- Hook prepends \`"crates/"\` before calling \`resolve-package-name\`: \`"crates/${SINGLE_CRATE_DIR}"\` — correct
- Fallback \`|| printf '%s' "$SINGLE_CRATE_DIR"\` fires when \`cargo xtask\` is unavailable (cold build, incomplete checkout). \`2>/dev/null\` suppresses stderr from cargo. Graceful degradation confirmed.
- \`SINGLE_CRATE_COUNT = "1"\` guard ensures the resolution is only attempted for single-crate pushes — correct scope

**Path normalization in \`resolve_package_names\`:**
- \`root_str.to_string_lossy()\` on Windows produces backslash paths (e.g., \`C:\project\`)
- Both \`manifest_normalized\` and \`root_normalized\` independently normalize \`\\\` → \`/\` before \`strip_prefix\`
- Root path has no trailing separator (confirmed: \`Path::to_string_lossy\` never adds one)
- Chain: \`strip_prefix(root)\` → \`strip_prefix('/')\` → \`strip_suffix("/Cargo.toml")\` is correct for paths like \`/project/crates/perl-lsp/Cargo.toml\`

**\`resolve_single_package_name\` logic:**
- Creates a single-element \`BTreeSet\`, calls \`resolve_package_names\`, takes \`.next()\` on the result
- For the matching case: exactly one element in result set, \`.next()\` returns it — correct
- For the non-matching case: empty set, \`.next()\` returns \`None\`, \`.ok_or_else\` returns descriptive error — correct
- \`trim_end_matches('/')\` handles trailing slashes (verified by test F)

**Integration test coverage:**
- 12 tests covering: subcommand help, dir/pkg mismatch (synthetic + real workspace), dir/pkg match, unknown dir, output format, trailing slash, multiple slashes, empty arg, workspace root, Windows compat, stdout cleanliness

### Edge cases examined

| Scenario | Status |
|----------|--------|
| crates/ directly in root (no sub-crate) | Safe: \`SINGLE_CRATE_ALL_UNDER_CRATES\` gate prevents empty SINGLE_CRATE_DIR reaching xtask |
| xtask not built yet | Safe: fallback to basename |
| Windows backslash in root path | Fixed: both sides normalized before strip_prefix |
| Trailing slash in crate_dir arg | Fixed: trim_end_matches('/') |
| Empty crate_dir | Fails resolution → non-zero exit (test H guards this) |
| Multiple packages mapping to same dir | Not possible: one Cargo.toml per directory |

### Fix forward
No logic bugs found. The implementation is correct. No changes made.

### Verdict
**APPROVED.** Logic is correct, edge cases are covered, tests are thorough.

---
*Reviewer deep — correctness gate passed.*

**Comment by EffortlessSteven** [comment]
## Green CI: SHA Verified

**HEAD SHA:** \`6059c466ce0ecc865ddce9523a7c5d87349f229f\`

| Check | Result |
|-------|--------|
| PR Smoke (Fast Feedback) | SUCCESS |
| UX Regression Tests | SUCCESS |
| validate-title | SUCCESS |
| GitGuardian Security Checks | SUCCESS |
| CI Gate (Merge-Blocking) | SKIPPED (expected on PR smoke pass) |

All CI checks green on current HEAD. No stale results.

---
*Green CI — SHA-verified fresh green.*

**Comment by EffortlessSteven** [comment]
## Diff Auditor: Final Coherence Check

**Verdict: CLEAN**

### Diff vs spec alignment
All 11 acceptance criteria addressed:
- \`cargo xtask resolve-package-name crates/perl-lsp\` → \`perl-lsp-rs\`: CLI subcommand in \`main.rs\` + \`targeted_checks.rs\`
- Pre-push hook calls \`resolve-package-name\`: \`hooks/pre-push\` change confirmed
- Graceful fallback: \`2>/dev/null || printf '%s' "$SINGLE_CRATE_DIR"\` present
- All 3 specified regression tests present + 6 edge case tests added by green-tdd
- Clippy clean (production code), formatting clean
- \`xtask/src/tasks/fmt.rs\` NOT changed (confirmed: 0 lines diff)

### Scope cleanliness
7 files changed — all within spec scope:
- \`.spec/4512-fmt-package-lookup/\` (3 planning docs — history, expected per CLAUDE.md)
- \`hooks/pre-push\` (in scope)
- \`xtask/src/main.rs\` (in scope)
- \`xtask/src/tasks/targeted_checks.rs\` (in scope)
- \`xtask/tests/fmt_package_lookup_tests.rs\` (in scope)

No unrelated files touched. No other crates modified.

### Leftover artifacts
None found. No TODO/FIXME/HACK/XXX/dbg!/eprintln! in diff.

### Commit coherence
```
328a5139a plan(xtask): spec files
b645535c2 test(xtask): red TDD (failing tests)
569883229 fix(hook): builder implementation
8be02c9ff test(xtask): green TDD (edge cases)
5227a03fa fix(xtask): bot comment cleanup (Result<TempDir>, spec doc fix)
6059c466c refactor(xtask): green-refactor (println style, pub(crate))
```
Clean story: plan → red → impl → green → review cleanup → refactor.

### Late commit regression check
12/12 tests pass on HEAD \`6059c466c\`. No regressions.

### PR metadata
- Title: \`fix(hook): resolve cargo package name from Cargo.toml, not dir basename (#4512)\` — ends with \`(#4512)\` ✓
- Labels: deep-reviewed, review-reviewed, maintainer-pr-reviewed, pr-responded, ci-green, green-refactor-reviewed, refactor-planner-reviewed ✓

Ready for merge.

---
*Diff auditor — CLEAN.*
